src/test/java/com/yahoo/bullet/pubsub/BufferingSubscriberTest.java

```diff
@@ -0,0 +1,241 @@
+/*
+ *  Copyright 2017, Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.pubsub;
+
+import lombok.Getter;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class BufferingSubscriberTest {
+    @Getter
+    private final static class ExampleBufferingSubscriber extends BufferingSubscriber {
+        private List<PubSubMessage> testMessages;
+        private int callCount = 0;
+
+        public ExampleBufferingSubscriber(int max, List<PubSubMessage> messages) {
+            super(max);
+            testMessages = new LinkedList<>();
+            add(messages);
+        }
+
+        public void add(List<PubSubMessage> messages) {
+            testMessages.addAll(messages);
+        }
+
+        @Override
+        public List<PubSubMessage> getMessages() {
+            callCount++;
+            if (testMessages == null || testMessages.isEmpty()) {
+                return testMessages;
+            }
+            return Collections.singletonList(testMessages.remove(0));
+        }
+
+        @Override
+        public void close() {
+            testMessages = new LinkedList<>();
```

Comment:
testMessages.clear() ?

Reply:
On purpose to test what haveMessages does when we return an empty list. See line 84 in this file

Reply:
Sorry misread your comment. Will fix.